dask/array/rechunk.py

```diff
@@ -169,31 +169,44 @@ def _old_to_new(old_chunks, new_chunks):
       [(2, slice(5, 10, None))],
       [(3, slice(0, 10, None)), (4, slice(0, 10, None))]]]
     """
-    old_known = [x for x in old_chunks if not any(math.isnan(y) for y in x)]
-    new_known = [x for x in new_chunks if not any(math.isnan(y) for y in x)]
 
-    n_missing = [sum(math.isnan(y) for y in x) for x in old_chunks]
-    n_missing2 = [sum(math.isnan(y) for y in x) for x in new_chunks]
+    def is_unknown(dim):
+        return any(math.isnan(chunk) for chunk in dim)
 
-    cmo = cumdims_label(old_known, "o")
-    cmn = cumdims_label(new_known, "n")
-
-    sums = [sum(o) for o in old_known]
-    sums2 = [sum(n) for n in new_known]
+    old_is_unknown = [is_unknown(dim) for dim in old_chunks]
+    new_is_unknown = [is_unknown(dim) for dim in new_chunks]
 
-    if not sums == sums2:
-        raise ValueError(f"Cannot change dimensions from {sums!r} to {sums2!r}")
-    if not n_missing == n_missing2:
+    if old_is_unknown != new_is_unknown or any(
+        new != old for new, old in compress(zip(old_chunks, new_chunks), old_is_unknown)
+    ):
```

Comment:
This is check subtly wrong, I think, depending on the provenance of the nans inserted in the chunk lists (and the interpreter being used). It happens to work because `normalize_chunks` normalises all nan entries to the singleton `np.nan`. However, it still feels a bit magic.

For example:
```python
import numpy as np
import math
[1, math.nan] != [1, np.nan] # => True
[1, math.nan] != [1, math.nan] # => False
[1, float("nan")] != [1, float("nan")] # => False
```

This is because cpython implements list equality not as (documented in https://docs.python.org/3/reference/expressions.html#value-comparisons)
```
# assumption lengths the same
all(a == b for a, b in zip(left, right))
```
But instead
```
# assumption all lengths same
all(a is b or a == b for a, b in zip(left, right)
```

Do we want to change to using a different sentinel for "missing value", perhaps `ellipsis` ?

Reply:
I guess we could explicitly unroll the logic that CPython implements for list equality if that doesn't make things impossible to comprehend?

Reply:
I suspect CPython is doing this in C s.t. our imitation would perform really badly. Not sure if this is a concern here. As I said, if this is not easy, I would prefer us not cleaning this up.

Reply:
@wence-: I like your suggestion for the documentation and have added it. Instead of preprocessing in various places, I'd rather think about how we can (and whether we want to) propagate this information through typing in the future.

Reply:
@wence-: Let me know if this works for you. If not, I'll just add your recommended preprocessing step, I don't have strong feelings here.

Reply:
Thanks, I agree that having a tighter specification of what a chunk specification is would be better (instead of preprocessing/massaging of unstructured data). But this documentation is good for now.